File: PracticeSort.java
import java.util.Arrays;
import java.util.Random;

public class PracticeSort {

    private static int[] assist;

    public static void main(String[] args) {
        int[] arr = { 8, 6, 6, 5, 4, 3, 2, 1, 0, 6 };
        System.out.println("排序前的数组：");
        System.out.println(Arrays.toString(arr));

        int[] arr1 = bubbleSort(arr);
        System.out.println("冒泡排序后的数组：");
        printArray(arr1);

        int[] arr2 = selectionSort(arr);
        System.out.println("选择排序后的数组：");
        printArray(arr2);

        int[] arr3 = insertionSort(arr);
        System.out.println("插入排序后的数组：");
        printArray(arr3);

        int[] arr4 = shellSort(arr);
        System.out.println("希尔排序后的数组：");
        printArray(arr4);

        int[] arr5 = mergeSort(arr);
        System.out.println("归并排序后的数组：");
        printArray(arr5);

        // int[] longArr = new int[100000];
        // for (int i = 0; i < longArr.length; i++) {
        // longArr[i] = 10000 - i;
        // }
        int[] longArr = generateRandomNumbers(10000);
        int[] sortedLongArr = longArr.clone();
        Arrays.sort(sortedLongArr);
        boolean result;

        long start = System.currentTimeMillis();
        result = Arrays.equals(insertionSort(longArr), sortedLongArr);
        long end = System.currentTimeMillis();
        System.out.println(result + " 程序运行时间：" + (end - start) + " 毫秒");

        start = System.currentTimeMillis();
        result = Arrays.equals(shellSort(longArr), sortedLongArr);
        end = System.currentTimeMillis();
        System.out.println(result + " 程序运行时间：" + (end - start) + " 毫秒");

        start = System.currentTimeMillis();
        result = Arrays.equals(mergeSort(longArr), sortedLongArr);
        end = System.currentTimeMillis();
        System.out.println(result + " 程序运行时间：" + (end - start) + " 毫秒");
    }

    public static int[] bubbleSort(int[] arr) {
        int[] array = arr.clone();
        int n = array.length;
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (array[j] > array[j + 1]) {
                    // 交换 array[j] 和 array[j+1]
                    int temp = array[j];
                    array[j] = array[j + 1];
                    array[j + 1] = temp;
                }
            }
        }
        return array;
    }

    public static int[] selectionSort(int[] arr) {
        int[] array = arr.clone();
        int N = array.length;
        int min_index;
        for (int i = 0; i < N - 1; i++) {
            min_index = i;
            for (int j = i + 1; j < N; j++) {
                if (array[j] < array[min_index]) {
                    min_index = j;
                }
            }
            int temp = array[min_index];
            array[min_index] = array[i];
            array[i] = temp;
        }
        return array;
    }

    public static int[] insertionSort(int[] arr) {
        int[] array = arr.clone();
        for (int i = 0; i < array.length - 1; i++) {
            for (int j = i + 1; j > 0; j--) {
                if (array[j - 1] > array[j]) {
                    int temp = array[j - 1];
                    array[j - 1] = array[j];
                    array[j] = temp;
                }
            }
        }
        return array;
    }

    public static int[] shellSort(int[] arr) {
        int[] array = arr.clone();
        int length = array.length;
        int gap = (length / 2);

        while (gap > 0) {
            for (int start = 0; start < gap; start++) {
                for (int i = start; i < length - gap; i = i + gap) {
                    for (int j = i + gap; j > start; j = j - gap) {
                        if (array[j - gap] > array[j]) {
                            int temp = array[j - gap];
                            array[j - gap] = array[j];
                            array[j] = temp;
                        } else {
                            break;
                        }
                    }
                }
            }
            gap = gap / 2;
        }
        return array;
    }

    public static int[] mergeSort(int[] arr) {
        int[] array = arr.clone();
        assist = new int[array.length];

        mergeSort_sort(array, 0, arr.length - 1);
        return array;
    }

    public static void mergeSort_sort(int[] array, int low, int high) {
        if (low > high) {
            return;
        }
        int mid = low + (high - low) / 2;
        if (!(low == mid)) {
            mergeSort_sort(array, low, mid);
        }
        if (!(mid + 1 == high)) {
            mergeSort_sort(array, mid + 1, high);
        }
        mergeSort_merge(array, low, mid, high);
    }

    public static void mergeSort_merge(int[] array, int low, int mid, int high) {
        int array1_low = low, array1_high = mid;
        int array2_low = mid + 1, array2_high = high;

        int p1 = array1_low;
        int p2 = array2_low;
        int p3 = low;

        while (true) {
            if (p1 > array1_high) {
                for (int i = 0; i <= high - p3; i++) {
                    assist[p3 + i] = array[p2 + i];
                }
                break;
            }
            if (p2 > array2_high) {
                for (int i = 0; i <= high - p3; i++) {
                    assist[p3 + i] = array[p1 + i];
                }
                break;
            }

            if (array[p2] < array[p1]) {
                assist[p3] = array[p2];
                p2 += 1;
            } else {
                assist[p3] = array[p1];
                p1 += 1;
            }
            p3 += 1;
        }

        for (int i = low; i <= high; i++) {
            array[i] = assist[i];
        }
    }

    public static void printArray(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static int[] generateRandomNumbers(int count) {
        int[] numbers = new int[count];
        Random random = new Random();

        for (int i = 0; i < count; i++) {
            numbers[i] = random.nextInt(10000);
        }

        return numbers;
    }
}

File: BasicPractice/PracticeSort.py
from copy import deepcopy
import random
from timeit import default_timer as timer
import argparse
from typing import List


def time(func):
    def wrapper(*args, **kwargs):
        start_time = timer()
        result = func(*args, **kwargs)
        end_time = timer()
        execution_time = end_time - start_time
        print(f"[INFO] {func.__name__} Running Time: {execution_time:.3f} seconds")
        return result

    return wrapper


@time
def bubble_sort(arr):
    array = deepcopy(arr)
    for i in range(len(array)):
        for j in range(0, len(array) - i - 1):
            if array[j] > array[j + 1]:
                array[j], array[j + 1] = array[j + 1], array[j]

    return array


@time
def selection_sort(arr):
    array = deepcopy(arr)
    for i in range(len(array) - 1):
        min_index = i
        for j in range(i + 1, len(array)):
            if array[min_index] > array[j]:
                min_index = j
        array[min_index], array[i] = array[i], array[min_index]
    return array


@time
def insertion_sort(arr):
    array = deepcopy(arr)
    for i in range(len(array) - 1):
        for j in range(i + 1, 0, -1):
            if array[j - 1] > array[j]:
                array[j - 1], array[j] = array[j], array[j - 1]
            else:
                break
    return array


@time
def shell_sort_stupid(arr):
    array = deepcopy(arr)
    gap = int(len(array) / 2)

    # 针对每个不为0的gap，分组排序
    while gap > 0:
        for start in range(0, gap):
            for i in range(start, len(array) - gap, gap):
                for j in range(i + gap, start, -gap):
                    if array[j - gap] > array[j]:
                        array[j - gap], array[j] = array[j], array[j - gap]
                    else:
                        break
        gap = int(gap / 2)
    return array


@time
def shell_sort_clever(arr):
    array = deepcopy(arr)
    gap = int(len(array) / 2)

    while gap > 0:
        for i in range(0, len(array) - gap):
            for j in range(i + gap, 0, -gap):
                if array[j - gap] > array[j]:
                    array[j - gap], array[j] = array[j], array[j - gap]
                else:
                    break
        gap = int(gap / 2)
    return array


@time
def merge_sort(arr):
    def sort(
        array: List,
        low: int,
        high: int,
    ):
        if low > high:
            return
        mid = low + (high - low) // 2
        if not low == mid:
            sort(array, low, mid)
        if not mid + 1 == high:
            sort(array, mid + 1, high)

        merge(array, low, mid, high)

    def merge(array: List, low: int, mid: int, high: int):
        array1_low, array1_high = low, mid
        array2_low, array2_high = mid + 1, high

        p1 = array1_low
        p2 = array2_low
        p3 = low

        while True:
            if p1 > array1_high:
                assist[p3 : high + 1] = array[p2 : array2_high + 1]
                break
            if p2 > array2_high:
                assist[p3 : high + 1] = array[p1 : array1_high + 1]
                break

            if array[p2] < array[p1]:
                assist[p3] = array[p2]
                p2 += 1
            else:
                assist[p3] = array[p1]
                p1 += 1
            p3 += 1

        array[low : high + 1] = assist[low : high + 1]

    array = deepcopy(arr)
    assist = [0] * len(array)

    sort(array, 0, len(array) - 1)

    return array


@time
def quick_sort(arr):
    def sort(array, low: int, high: int):
        if low >= high:
            return
        partitionIndex = partition(array, low=low, high=high)
        sort(array, low, partitionIndex - 1)
        sort(array, partitionIndex + 1, high)

    def partition(array, low: int, high: int):
        pivot = array[low]
        left = low + 1
        right = high

        while True:
            if array[left] <= pivot:
                if left == high:
                    break
                left += 1
                continue

            if array[right] > pivot:
                if right == low:
                    break
                right -= 1
                continue

            if right < left:
                break

            array[left], array[right] = array[right], array[left]

        array[low], array[right] = array[right], array[low]
        return right

    array = deepcopy(arr)

    sort(array, 0, len(array) - 1)

    return array


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Sort")
    parser.add_argument(
        "-t",
        "--timeit",
        action="store_true",
        required=False,
        default=False,
        help="是否用大样本计时",
    )
    args = parser.parse_args()

    if not args.timeit:
        arr = [8, 6, 6, 5, 4, 3, 2, 1, 0, 6]
    else:
        arr = [random.randint(0, 10000) for _ in range(0, 10000)]

    arr0 = sorted(arr)
    arr1 = bubble_sort(arr)
    arr2 = selection_sort(arr)
    arr3 = insertion_sort(arr)
    arr4 = shell_sort_stupid(arr)
    arr5 = shell_sort_clever(arr)
    arr6 = merge_sort(arr)
    arr7 = quick_sort(arr)

    if args.timeit:
        print(f"冒泡排序: {arr1 == arr0}")
        print(f"选择排序: {arr2 == arr0}")
        print(f"插入排序: {arr3 == arr0}")
        print(f"希尔排序(stupid): {arr4 == arr0}")
        print(f"希尔排序(clever): {arr5 == arr0}")
        print(f"归并排序: {arr6 == arr0}")
        print(f"快速排序: {arr7 == arr0}")
    else:
        print(f"原始序列: {arr}")
        print(f"官方排序: {arr0}")
        print(f"冒泡排序: {arr1}")
        print(f"选择排序: {arr2}")
        print(f"插入排序: {arr3}")
        print(f"希尔排序(stupid): {arr4}")
        print(f"希尔排序(clever): {arr5}")
        print(f"归并排序: {arr6}")
        print(f"快速排序: {arr7}")
